### ISSUE TITLE
JSON: preserve raw string for ints and floats

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#2506]
Right now if a float/int has a lot of digits some precision is lost:

```crystal
require "json"

class Obj
  JSON.mapping({
    float: Float64,
  })
end

obj = Obj.from_json(%({"float": 1234.567891011121314}))
obj.float # => 1234.57
sprintf("%.16f", obj.float) # => "1234.5678910111212190"
```

We should probably retain the original string and make it accessible in JSON's lexer (via a raw property). Then a mapping could use it:

```crystal
class Obj
  JSON.mapping({
    float: {type: String, raw: true}, # use raw value, don't parse as float
  })
end
```

We could also do, once we have BigFloat:

```crystal
class Obj
  JSON.mapping({
    float: BigFloat,
  })
end

class BigFloat
  def self.new(pull : JSON::PullParser)
    # don't use the parsed float value, use the raw string
    new(pull.raw)
  end
end
```

Keeping the raw value will slow down JSON parsing performance a bit, but more real-world cases could be supported in this way, so this should be implemented.

**Comment by jhass** [comment]
I would love to see this generically, not just limited to integers and floats, but objects, arrays and null too. That would make working around previously mentioned issues like parsing partially unknown JSON or deciding a specific mapping based on a property on an outer object way easier (and probably faster), since we then no longer need to fallback to using the pull parser by hand or mapping into `JSON::Any`, serializing that again and parsing it with a second mapping once determinable.

**Comment by asterite** [comment]
@jhass Good idea, I'll try to implement that too.

I'm trying to remember if there are cases where the discriminator was inside the object to be created. For example ("kind" is the discriminator):

```json
{
  "location": {
    "kind": "Point",
    "x": 1,
    "y": 2
  }
}
```

And another:

```json
{
  "location": {
    "kind": "Line",
    "x1": 1,
    "y1": 2,
    "x2": 3,
    "y3": 4
  }
}
```

If so, it would be a good idea to make `JSON.mapping` also be able to deserialize an object from a `JSON::Any`:

```crystal
class SomeObject
  JSON.mapping({
    any_location: {type: JSON::Any, key: "location"},
  })

  def location
    @location ||= begin
      case any_location["kind"]
      when "Point"
        Point.from_json(any_location) # instead of Point.from_json(any_location.to_json)
      when "Line"
        Line.from_json(any_location)
      else
        raise "unknown kind"
      end
    end
  end
end
```

I think you proposed that in the past and I rejected it, but now I think it's useful.

**Comment by asterite** [comment]
For example there's [GeoJSON](https://en.wikipedia.org/wiki/GeoJSON):

```json
{ "type": "FeatureCollection",
    "features": [
      { "type": "Feature",
        "geometry": {"type": "Point", "coordinates": [102.0, 0.5]},
        "properties": {"prop0": "value0"}
        },
      { "type": "Feature",
        "geometry": {
          "type": "LineString",
          "coordinates": [
            [102.0, 0.0], [103.0, 1.0], [104.0, 0.0], [105.0, 1.0]
            ]
          },
        "properties": {
          "prop0": "value0",
          "prop1": 0.0
          }
        },
      { "type": "Feature",
         "geometry": {
           "type": "Polygon",
           "coordinates": [
             [ [100.0, 0.0], [101.0, 0.0], [101.0, 1.0],
               [100.0, 1.0], [100.0, 0.0] ]
             ]
         },
         "properties": {
           "prop0": "value0",
           "prop1": {"this": "that"}
           }
         }
       ]
     }
```

We can store the raw string for "coordinates" (might be nil) and then based on the "type" property we create them... though maybe mapping the whole object to JSON::Any and then, based on the "type" property, create different objects might be better.

**Comment by asterite** [comment]
Here's a program to parse that GeoJSON file :-)

```crystal
require "json"

alias PointTuple = Tuple(Float64, Float64)

class FeatureCollection
  JSON.mapping({
    type:     String,
    features: Array(Feature),
  })
end

class Feature
  JSON.mapping({
    type:       String,
    geometry:   Geometry,
    properties: {type: Hash(String, JSON::Any), nilable: true},
  })
end

abstract class Geometry
  def self.new(pull : JSON::PullParser)
    type = nil
    coordinates = nil

    pull.read_object do |key|
      case key
      when "type"
        type = pull.read_string
      when "coordinates"
        coordinates = pull.read_raw
      else
        pull.skip
      end
    end

    type = type.not_nil!
    coordinates = coordinates.not_nil!

    case type
    when "Point"
      Point.new(PointTuple.from_json(coordinates))
    when "LineString"
      LineString.new(Array(PointTuple).from_json(coordinates))
    when "Polygon"
      Polygon.new(Array(Array(PointTuple)).from_json(coordinates))
    else
      raise "unknown geometry type: #{type}"
    end
  end
end

class Point < Geometry
  JSON.mapping({
    type:        String,
    coordinates: PointTuple,
  })

  def initialize(@coordinates : PointTuple)
    @type = "Point"
  end
end

class LineString < Geometry
  JSON.mapping({
    type:        String,
    coordinates: Array(PointTuple),
  })

  def initialize(@coordinates : Array(PointTuple))
    @type = "LineString"
  end
end

class Polygon < Geometry
  JSON.mapping({
    type:        String,
    coordinates: Array(Array(PointTuple)),
  })

  def initialize(@coordinates : Array(Array(PointTuple)))
    @type = "Polygon"
  end
end

collection = FeatureCollection.from_json(File.read("./file.json"))
p collection
```

I didn't use `JSON.mapping` for `Geometry` but instead I used a pull parser. I had to read the raw value for the coordinates and then parse it according to the geometry type. I guess we won't need to parse an object from JSON::Any, at least in this example parsing it from the raw string is enough.

**Comment by bcardiff** [comment]
I see there is a value in preserving the raw information. For tracing back errors mainly. But at the same time it force to keep the whole string in memory.

I don't like the idea of re-iterating again the json string, despite the way it will be parsed. It kind of defeat the purpose of JSON.mapping.

I like the idea of having (optionally) a raw JSON::Any value, despite the formatter used, at the end of the day plain JSON is the underlying format. And for sure the support for custom type based on a discriminator would be great.

And as a side note, I saw many times the `Hash(String, JSON::Any)`, couldn't we have a `JSON::Record` alias for that?

**Comment by asterite** [comment]
@bcardiff I also thought in the beginning that the whole string was going to be kept in memory, but I found a way not to need that. The raw string is only kept for ints and floats. To get the raw string of a big object, PullParser recreates it as it parses the object. This can be seen [here](https://github.com/crystal-lang/crystal/blob/feature/json_raw/src/json/pull_parser.cr#L119-L181). Recreating this string is only needed when invoking `read_raw` so performance isn't lost when this method is not invoked (well, just a bit because the original string for ints and floats is always preserved). And recreating it only involves at most one allocation (`String.build { ... }`) and some reallocations of the string buffer. This can be improved in the future a bit by not always generating this string... but right now I'm lazy (I checked kostya's benchmarks and they get just a bit slower, not too much)

I prefer to have this functionality as it will allow many more uses cases, and later optimize it without changing the API (I think it's possible)

`JSON::Record` sounds good, I don't know about the name. Maybe `JSON::Bag`? Or maybe having it explicit as a Hash is better.
